addons/mail/models/res_users.py

```diff
@@ -57,6 +57,12 @@ def _compute_moderation_counter(self):
         for user in self:
             user.moderation_counter = result.get(user.id, 0)
 
+    @api.constrains('notification_type')
+    def _check_notification_type_public_users(self):
+        for user in self:
+            if user.share and user.notification_type == 'inbox':
+                raise exceptions.ValidationError(_("A non internal user can only have their notifications handled by email"))
```

Comment:
We could be a bit more specific by adding the name, like "User %(name) is a portal or public user and can therefore only have its notifications handled by email."